source/language/gates.rst

```diff
@@ -164,23 +283,140 @@ of the gate definition.
-statements are valid.
+'General' gate declarations have a similar structure, with minor differences:
 
-The statements in the body
-can only refer to the symbols given in the parameter or argument list,
-and these symbols are scoped only to the subroutine body.
+.. code-block:: c
 
-An empty body corresponds to the identity gate.
+   // comment
+   gate name(typedParams) : typedQargs
+   {
+     body
+   }
+
+The (again, optional) parameter list ``typedParams`` is a comma-separated list of variable
+parameters, which in this case must be provided with explicit type specifications.
+The list ``typedQargs`` is a comma-separated list of operands which are either individual
+qubits, or registers/arrays of qubits, where each is again provided with an explicit type
+specification. 
+If there are no variable parameters, the parentheses are optional. At least
+one quantum operand is required, and the quantum operands
+must be immediately preceded by a ``:`` delimiter.
+The arguments in ``typedQargs`` can be indexed within the body
+of the gate definition if, and only if, it is a register or array of qubits.
+
+.. code-block:: c
+
+   // this is ok:
+   gate g(angle alpha, int k): qubit[3] a
+   {
+     U(0, 0, alpha) a[0];
+     U(0, 0, alpha/k) a[1];
+     U(0, 0, alpha/(k**2)) a[2];
+   }
+
+   // this is also ok:
+   gate qutritX : qubit[2] a {
+     x a[1];
+     cx a[1], a[0];
+     cx a[0], a[1];
+   }
+
+   // this is invalid:
+   gate g(angle alpha, int k): qubit a
+   {
+     U(0, 0, alpha) a[0]; // not allowed to index an individual qubit operand
+     U(0, 0, alpha/k) a[1];
+     U(0, 0, alpha/(k**2)) a[2];
+   }
+
+
+For either kind of gate declaration, the ``body`` may consist of the following:
+
+ * declaration and initialisation of classical variables (but not re-assignment to them);
+
+ * program logic (``if`` statements and ``for`` loops) with conditions/bounds involving constants,
+   and simple expressions depending on the gate arguments and local identifiers / loop iterators;
+
+ * and calls to built-in gates and previously defined gates.
+
+For instance, the following defines a quantum Fourier transform on eight qubits:
 
-To avoid infinite recursion, gates must be declared before use and
+.. code-block:: c
+  gate QFT256 : qubit[8] q {
+        uint n = 8;
+        for uini j in [0 : n-1] {
+          h q[j];
+          for uint k in [j+1 : n-1]
+           ctrl @ Rz(pi / 2**(k-j+1)) q[j], q[k];
+          }
+        for uint j in [0:n-1]
+          if (j != n-1-j)
+            swap q[j], q[n-1-j]
+  }
+
+Classical storage types and parameters in a ``gate`` body are treated as being immutable,
+and cannot be assigned to more than once. (For loop induction variables are treated as being constant within
+the scope of any single iteration of the ``for`` loop, and can only be modified by the logic of the loop
+itself.) External functions, ``reset`` operations, ``measure`` operations (or other operations with potentially
```

Comment:
```suggestion
itself.) ``extern`` functions, ``reset`` operations, ``measure`` operations (or other operations with potentially
```
Probably better to say `extern` rather than "External".